### PR TITLE
Added migration to populate cancellation events

### PR DESCRIPTION
**Pull request by allouis**
refs https://github.com/TryGhost/Team/issues/1302

This also includes the MRR delta changes to the events, which may need
to be split out into a separate migration.

### CODE REVIEW

File: core/server/data/migrations/versions/4.43/2022-04-07-11-18-add-cancelled-events.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+        const eventsToInsert = cancelledSubscriptions.map((subscription) => {
+            const event = {
+                id: (new ObjectID()).toHexString(),
+                type: 'cancelled',
+                member_id: subscription.id,
+                from_plan: null,
+                to_plan: null,
+                currency: subscription.plan_currency,
+                source: 'migration',
```
Comment (allouis) [inline core/server/data/migrations/versions/4.43/2022-04-07-11-18-add-cancelled-events.js:33]: Is this a good idea @matthanley? Should we use this as a constraint for the `down` migration? 🤔

File: core/server/data/migrations/versions/4.43/2022-04-08-11-54-add-cancelled-events.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+        if (cancelledSubscriptions.length === 0) {
+            logging.info('No missing cancelled events - skipping migration');
+            return;
+        }
+
+        const eventsToInsert = cancelledSubscriptions.map((subscription) => {
+            const event = {
+                id: (new ObjectID()).toHexString(),
+                type: 'cancelled',
```
[the remark this answers is not in the document]
Comment (daniellockyer) [reply core/server/data/migrations/versions/4.43/2022-04-08-11-54-add-cancelled-events.js:29]: vs above 😅 

```
.whereNot('members_stripe_customers_subscriptions.status', 'canceled');
```

[the remark this answers is not in the document]
Comment (matthanley) [reply core/server/data/migrations/versions/4.43/2022-04-08-11-54-add-cancelled-events.js:29]: We have a precedent in the db from the Stripe subs, I think we should follow it for consistency. We also use US-English everywhere, so I think single-L makes sense.

File: core/server/data/migrations/versions/4.43/2022-04-07-11-18-add-cancelled-events.js (hunk 1 above)
Comment (matthanley) [reply core/server/data/migrations/versions/4.43/2022-04-07-11-18-add-cancelled-events.js]: Yeah I think this probably makes sense. We can always update it later if we need to.

**Review by matthanley — CHANGES_REQUESTED** [review]

**Comment by allouis** [comment]
@matthanley Updated spelling!

File: core/server/data/migrations/versions/4.43/2022-04-08-11-54-add-cancelled-events.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+            return event;
+        });
+
+        logging.info(`Found ${eventsToInsert.length} missing cancellation events`);
+        await knex('members_paid_subscription_events').insert(eventsToInsert);
+    },
+    async function down(knex) {
+        logging.info('Deleting all members_paid_subscription_events with a "type" of "cancelled"');
+        await knex('members_paid_subscription_events').where('type', 'canceled').del();
```
Comment (matthanley) [inline core/server/data/migrations/versions/4.43/2022-04-08-11-54-add-cancelled-events.js:48]: Needs an additional check on `source=migration`

File: core/server/data/migrations/versions/4.43/2022-04-08-11-54-add-cancelled-events.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        }
+
+        const eventsToInsert = cancelledSubscriptions.map((subscription) => {
+            const event = {
+                id: (new ObjectID()).toHexString(),
+                type: 'canceled',
+                member_id: subscription.member_id,
+                subscription_id: subscription.id,
+                from_plan: null,
+                to_plan: null,
```
Comment (matthanley) [inline core/server/data/migrations/versions/4.43/2022-04-08-11-54-add-cancelled-events.js:33]: Rethinking this: will we be able to answer the question later on of "what plan were they on when they cancelled?" 🤔

**Review by matthanley — CHANGES_REQUESTED** [review]

File: core/server/data/migrations/versions/4.43/2022-04-08-11-54-add-cancelled-events.js (hunk 4 above)
Comment (allouis) [reply core/server/data/migrations/versions/4.43/2022-04-08-11-54-add-cancelled-events.js:33]: 🤔 Hmm - no - I guess we should store it as `from_plan` and `to_plan` ?

Comment (matthanley) [reply core/server/data/migrations/versions/4.43/2022-04-08-11-54-add-cancelled-events.js:33]: Yeah I think that makes sense

Comment (allouis) [reply core/server/data/migrations/versions/4.43/2022-04-08-11-54-add-cancelled-events.js:33]: Updated!

**Comment by allouis** [comment]
@daniellockyer This is ready for you now!

**Review by daniellockyer — APPROVED** [review]

**Review by matthanley — APPROVED** [review]

**Pull request merged**